plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs

```diff
@@ -0,0 +1,120 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE TypeFamilies      #-}
+module Language.Plutus.Contract.Contract(
+      Contract
+    , done
+    , emit
+    , waiting
+    , applyInput
+    , applyInputs
+    , drain
+    , outputs
+    , isDone
+    , await
+    , result
+    , loopM
+    ) where
+
+import           Control.Applicative (liftA2)
+import           Control.Monad       ((>=>))
+import           Data.Bifunctor      (first)
+import           Data.List           (foldl')
+
+data Contract t i a =
+    Waiting (i -> Contract t i a)
+    | Done
+    | Emit t (Contract t i a) -- produce a 't' value
+    | Pure a
+    deriving (Functor)
+
+-- The applicative instance parallelises the 'Waiting' operations
+instance Applicative (Contract t i) where
```

Comment:
c.f. http://hackage.haskell.org/package/haxl-2.0.1.1/docs/src/Haxl.Core.Monad.html#line-569

Haxl is also doing the parallel applicative thing, and it has a similar structure.

Reply:
Yes I looked at haxl, but it's tied closely to the IO monad and geared towards concurrent execution and caching, which we don't need. The `concur` package also goes in a similar direction.

Reply:
Yeah, I just meant we can cross-check our implementation of stuff.